packages/realm/src/schema/normalize.ts

```diff
@@ -387,16 +388,39 @@ function assertNotUsingShorthand(input: string | undefined, info: PropertyInfo):
 /**
  * Get an error message for an invalid property schema.
  */
-function errMessage({ objectName, propertyName }: PropertyInfo, message: string): string {
+function propSchemaErrorMessage({ objectName, propertyName }: PropertyInfo, message: string): string {
   return `Invalid type declaration for property '${propertyName}' on '${objectName}': ${message}`;
 }
 
+/**
+ * Generate an error caused by an invalid property schema.
+ * (Returning a function rather than the Error itself in order
+ * for the Error to only be created if needed.)
+ */
+function propError(info: PropertyInfo, message: string): () => SchemaParseError {
+  return () => new SchemaParseError(propSchemaErrorMessage(info, message));
+}
+
+/**
+ * Get an error message for an invalid object schema.
+ */
+function objectSchemaErrorMessage(objectName: string, message: string): string {
+  const displayName = objectName ? `object '${objectName}'` : "unnamed object";
+  return `Invalid schema for ${displayName}: ${message}`;
+}
+
+/**
+ * Generate an error caused by an invalid object schema.
+ */
+function objectError(objectName: string, message: string): () => SchemaParseError {
+  return () => new SchemaParseError(objectSchemaErrorMessage(objectName, message));
+}
+
```

Comment:
Did you consider making each of these cases subclasses the `SchemaParseError` instead?

Reply:
Good idea!

We could have these:

```ts
export class SchemaParseError extends Error {
  /** @internal */
  constructor(message: string) {
    super(message);
  }
}

export class ObjectSchemaParseError extends SchemaParseError {
  /** @internal */
  constructor(objectName: string, message: string) {
    const displayName = objectName ? `object '${objectName}'` : "unnamed object";
    super(`Invalid schema for ${displayName}: ${message}`);
  }
}

export class PropertySchemaParseError extends SchemaParseError {
  /** @internal */
  constructor(objectName: string, propertyName: string, message: string) {
    super(`Invalid type declaration for property '${propertyName}' on '${objectName}': ${message}`);
  }
}
```

Having the specific error class be responsible for the message formatting can be useful here. In that case, I could try to modify the [validation part](https://github.com/realm/realm-js/blob/4fb7b0f3c4d79da4a7e3360e719e8e9b2d01b7bb/packages/realm/src/schema/validate.ts#L128) to also throw the more specific one.